### PR TITLE
Avoid specialization in the metadata serialization code

### PR DESCRIPTION
**Pull request by bjorn3**
With the exception of a perf-only specialization for byte slices and byte vectors.

This uses the same trick of introducing a new trait and having the Encodable and Decodable derives add a bound to it as used for TyEncoder/TyDecoder. The new code is clearer about which encoder/decoder uses which impl and it reduces the dependency of rustc on specialization, making it easier to remove support for specialization entirely or turn it into a construct that is only allowed for perf optimizations if we decide to do this.

### CODE REVIEW

**Comment by bjorn3** [comment]
@rust-timer check 2230d934cfc30f77144eb4fc0c9fe8b8adf40c44

**Comment by bjorn3** [comment]
@rust-timer build 2230d934cfc30f77144eb4fc0c9fe8b8adf40c44

**Comment by Nilstrieb** [comment]
very cool. I'm wondering whether it makes sense to have the extra `Span{En,De}coder` traits or whether it may make more sense to just merge `rustc_serialize` and `rustc_span` and unify it into a single `{En,De}coder` trait...

**Comment by bjorn3** [comment]
@Nilstrieb I've considered that, but assumed it would be more controversial.

**Comment by Nilstrieb** [comment]
Let's keep it like this for now, we can always change it later.

File: compiler/rustc_span/src/lib.rs [hunk 1]
```diff
@@ -1016,22 +1019,43 @@ impl Default for Span {
     }
 }
 
-impl<E: Encoder> Encodable<E> for Span {
-    default fn encode(&self, s: &mut E) {
-        let span = self.data();
-        span.lo.encode(s);
-        span.hi.encode(s);
+pub trait SpanEncoder: Encoder {
```
Comment (Nilstrieb) [inline compiler/rustc_span/src/lib.rs:1028]: I think it would be nice if there were some doc comments here saying that this is used for why and it exists.

Comment (bjorn3) [reply compiler/rustc_span/src/lib.rs:1030]: Done

**Review by wesleywiser — APPROVED** [review]

File: compiler/rustc_span/src/lib.rs [hunk 2]
```diff
@@ -35,6 +35,8 @@
 #![feature(rustdoc_internals)]
 // tidy-alphabetical-end
 
+extern crate self as rustc_span;
+
```
[the remark this answers is not in the document]
Comment (bjorn3) [reply compiler/rustc_span/src/lib.rs:39]: The `Encodable` and `Decodable` derives reference `rustc_span` for the `SpanEncoder` and `SpanDecoder` trait bounds.

**Pull request merged**